src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -97,45 +109,192 @@ const std::shared_ptr<AbstractLQPNode>& SQLPipelineStatement::get_unoptimized_lo
   return _unoptimized_logical_plan;
 }
 
+bool SQLPipelineStatement::is_uniformly_distributed(const float distribution_threshold) {
+  auto& unoptimized_lqp = get_unoptimized_logical_plan();
+
+  const auto started_uniform_check = std::chrono::high_resolution_clock::now();
+
+  // find used columns
+  std::vector<std::shared_ptr<LQPColumnExpression>> column_expressions;
+  visit_lqp(unoptimized_lqp, [&column_expressions](const auto& node) {
+    if (node) {
+      for (auto& root_expression : node->node_expressions) {
+        visit_expression(root_expression, [&column_expressions](auto& expression) {
+          if (expression->type == ExpressionType::LQPColumn) {
+            auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+            column_expressions.push_back(column_expression);
+          }
+          return ExpressionVisitation::VisitArguments;
+        });
+      }
+    }
+    return LQPVisitation::VisitInputs;
+  });
+
+  bool contains_uniform_distribution = true;
+
+  // find used tables and check if used columns are uniformly distributed
+  visit_lqp(unoptimized_lqp, [&contains_uniform_distribution, &column_expressions, &distribution_threshold](const auto& node) {
+    if (node) {
+      const auto &table_node = std::dynamic_pointer_cast<StoredTableNode>(node);
```

Comment:
Hm, I am unsure about this, but I think in most other places we check the node's type and cast accordingly, don't we? That way we could also use a guard clause here.

Reply:
done